source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,166 @@
+//
+// ---------------------------------------------------------------------
+
+
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/point.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_tools.h>
+
+#ifdef DEAL_II_WITH_P4EST
+
+#include <vector>
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace GridTools
+  {
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim> >
+    compute_locally_owned_bounding_box
+    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
+    {
+      // In order to return the boundary points of the locally_owned cells correctly
+      // we need to handle the possibility of a second connected component.
+      //
+      // Assuming the triangulation is distributed following the morton-z curve order
+      // we know that, if there are two connected components, then the first and the
+      // last locally owned cell belong to different components.
+      //
+      // Thus we begin by looking for those cells which we call firstc and lastc
+
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      firstc = distributed_tria.begin_active();
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      lastc = distributed_tria.last_active();
+
+      while ( ! firstc->is_locally_owned() )
+        ++firstc;
+
+
+      while ( ! lastc->is_locally_owned() )
+        --lastc;
+
+      //Degenerate cases: only one owned cell or we're at level 0
+      if (lastc==firstc || firstc->level()*lastc->level()==0)
+        {
+          Point<spacedim> minp = firstc->center();
+          Point<spacedim> maxp = firstc->center();
+          for (unsigned int v=0; v<GeometryInfo<spacedim>::vertices_per_cell; ++v)
+            {
+              for ( unsigned int d=0; d<spacedim; ++d)
+                {
+                  minp[d] = std::min( minp[d], firstc->vertex(v)[d]);
+                  maxp[d] = std::max( maxp[d], firstc->vertex(v)[d]);
+                }
+            }
+          std::vector<BoundingBox<spacedim>> vBBox;
+          vBBox.push_back(BoundingBox<spacedim>(std::make_pair( minp, maxp )));
+          return vBBox;
+        }
+
+      //Now we look for parentc: the coarsest cell containing both firstc and lastc
+      //This is the coarsest level at which we can separate the connected components
+      Point<spacedim> lastc_center = lastc->center();
+      typename parallel::distributed::Triangulation< dim, spacedim >::cell_iterator
+      parentc = firstc;
+      //tmpc shall contain the children of parentc from which firstc descends
+      typename parallel::distributed::Triangulation< dim, spacedim >::cell_iterator
+      tmpc = firstc;
+
+      while (parentc->level()>0)
+        {
+          parentc = tmpc->parent();
+          if (parentc->point_inside(lastc_center))
```

Comment:
I am not sure if this is robust enough (I don't quite understand the logic). There are cases where the center of a child cell is not inside all parents (think refining curved geometry).